src/v/datalake/record_translator.cc

```diff
@@ -0,0 +1,113 @@
+#include "iceberg/datatypes.h"
+#include "iceberg/values.h"
+#include "iceberg/values_avro.h"
+
+#include <avro/Generic.hh>
+#include <avro/GenericDatum.hh>
+
+namespace datalake {
+
+namespace {
+
+struct value_translating_visitor {
+    // Buffer ready to be parsed, e.g. no schema ID or protobuf offsets.
+    iobuf parsable_buf;
+    const iceberg::field_type& type;
+
+    ss::future<optional_value_outcome>
+    operator()(const google::protobuf::Descriptor& d) {
+        return deserialize_protobuf(std::move(parsable_buf), d);
+    }
+    ss::future<optional_value_outcome> operator()(const avro::ValidSchema& s) {
+        avro::GenericDatum d(s);
+        try {
+            auto in = std::make_unique<iceberg::avro_iobuf_istream>(
+              std::move(parsable_buf));
+            auto decoder = avro::validatingDecoder(s, avro::binaryDecoder());
+            decoder->init(*in);
+            avro::decode(*decoder, d);
+        } catch (...) {
+            co_return value_conversion_exception(fmt::format(
+              "Error reading Avro buffer: {}", std::current_exception()));
+        }
+        co_return iceberg::val_from_avro(d, type, iceberg::field_required::yes);
+    }
+};
+
+} // namespace
+
+record_type
+record_translator::build_type(std::optional<resolved_type> val_type) {
+    auto ret_type = schemaless_struct_type();
+    std::optional<schema_identifier> val_id;
+    if (val_type.has_value()) {
+        // Add the extra user-defined fields.
+        ret_type.fields.emplace_back(iceberg::nested_field::create(
+          0,
+          val_type->type_name,
+          iceberg::field_required::no,
+          std::move(val_type->type)));
+        val_id = std::move(val_type->id);
+    }
+    return record_type{
+      .comps = record_schema_components{
+          .key_identifier = std::nullopt,
+          .val_identifier = std::move(val_id),
+      },
+      .type = std::move(ret_type),
+    };
+}
+
+ss::future<checked<iceberg::struct_value, record_translator::errc>>
+record_translator::translate_data(
+  kafka::offset o,
+  iobuf key,
+  const std::optional<resolved_type>& val_type,
+  iobuf parsable_val,
+  model::timestamp ts) {
+    auto ret_data = iceberg::struct_value{};
+    ret_data.fields.emplace_back(iceberg::long_value(o));
+    // NOTE: Kafka uses milliseconds, Iceberg uses microseconds.
+    ret_data.fields.emplace_back(iceberg::timestamp_value(ts.value() * 1000));
+    ret_data.fields.emplace_back(iceberg::binary_value{std::move(key)});
+    if (val_type.has_value()) {
+        // Fill in the internal value field.
+        ret_data.fields.emplace_back(std::nullopt);
+
+        auto translated_val = co_await std::visit(
+          value_translating_visitor{std::move(parsable_val), val_type->type},
+          val_type->schema);
+        if (translated_val.has_error()) {
+            vlog(
+              datalake_log.error,
+              "Error converting buffer: {}",
+              translated_val.error());
```

Comment:
is this the case for a DLQ? in think in such case we could probably also pass this as an unparsed binary value..

(unrelated, I was also wondering if there is any use to having an optional schema_id in the default schema for debugging, in such a case one could look up the schema by id and attempt to debug the unparsed value)

Reply:
I don't know that this would necessarily be _where_ the DLQ would be, but this _is_ the case that would lead to a DLQ. I'm thinking we'd route the record at a higher level, rather than returning something here

Reply:
>(unrelated, I was also wondering if there is any use to having an optional schema_id in the default schema for debugging, in such a case one could look up the schema by id and attempt to debug the unparsed value)

Yeah, definitely (IMO). I think there are a few other fields we should add (partition id, maybe protobuf offsets). I opted to keep this PR focused on the mechanisms, and am expecting us to finalize the default fields in the next couple weeks

Reply:
> but this is the case that would lead to a DLQ.

cool thats what I meant